### DISCUSSION TITLE
Some Issues Implementing the Debug Renderer

### DISCUSSION THREAD

**Discussion by cuppajoeman** [jrouwe/JoltPhysics#964] [Q&A] [answered]
Hi there, 

I'm working on a game in opengl3.3 and I've been trying to implement the debug renderer so that I can see what jolt sees with respect to my game. 

So far I've built up a project where a player can move around a basic world aim move and jump using Jolt, right now all the rendering is being done manually, and just so happens to line up with what jolt sees for collisions etc...

https://github.com/jrouwe/JoltPhysics/assets/19198104/4b4d083f-4d3a-41f4-bbc0-dbaf30dfc0c1

I'm pretty new to c++ and I tried my best to grasp how `DebugRendererImp.cpp/hpp` works, and to implement that with respect to opengl, the result I got so far was this (note that the debug drawing is done in wireframe mode in grey)

https://github.com/jrouwe/JoltPhysics/assets/19198104/2b423113-837a-494e-a6a6-0a87e8170816

What I did was as follows:
* Instead of `BatchImpl` I created [`TriangleData`](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/c25f6898390d7b48daef7b87b87c6722442a21c8/graphics/physics_debug_renderer/physics_debug_renderer.hpp#L61), instead of being a render primitive it simply holds the triangle's data for later use. Note that I had to create the class `ThatIHaveToMake` to mirror the structure in the `BatchImpl` or else I couldn't get it to work.
* I implemented the function [`DrawGeometry`](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/c25f6898390d7b48daef7b87b87c6722442a21c8/graphics/physics_debug_renderer/physics_debug_renderer.cpp#L59C1-L59C40) in my class, which given some geometry attempts to render the triangles with opengl.

The result, as seen in my previous video is a bunch of crazy triangles, which seem to be changing all the time.

I would be very grateful if anyone could take a look over my code in [`physics_debug_renderer.cpp`](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/c25f6898390d7b48daef7b87b87c6722442a21c8/graphics/physics_debug_renderer/physics_debug_renderer.cpp) to provide any hints on how I can debug this, or if there are any glaring issues.

-----

Notes:
* The representation of the world's `MeshShape` is correct, I know this to be true because when moving around the world collisions occur correctly, therefore the data being stored by Jolt with respect to the map's `MeshShape` is correct.
* I have verified that the code in `DrawGeometry` can correctly render triangle data, for example by instead using a single triangle it correctly renders that in space, this means that the issue is with the data being stored in the `triangleBatches` or how I'm parsing it so I'm quite certain the issue is within my `CreateTriangleBatch` method.
* The debug renderer is able to render the wireframe of the stone walls on the sides perfectly, but all the other things seem very garbled, so at least one `TriangleData` gets created correctly by the `CreateTriangleBatch` function.
* No calls to `DrawLine` were being made so I know the issue is in `DrawGeometry`
* No calls to `CreateTriangleBatch` where the arguments are vertices along with indices occur, so I believe the issue is in `CreateTriangleBatch(const Triangle *inTriangles, int inTriangleCount)`
* My game doesn't use a LOD system yet, I don't fully understand how I'm supposed to use the LOD array. 
* Sometimes when I try to [load a triangle batch](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/c25f6898390d7b48daef7b87b87c6722442a21c8/graphics/physics_debug_renderer/physics_debug_renderer.cpp#L71C9-L71C61) I get back a null or 0x20 pointer, which I don't fully understand
* I find it very strange that the triangle data seems to be changing (causing the flickering effect), I have no idea how/why that could happen

**Comment by jrouwe** [comment, marked as answer]
I think the biggest problem is that you're storing a pointer to the triangle data that you get passed in:

https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/c25f6898390d7b48daef7b87b87c6722442a21c8/graphics/physics_debug_renderer/physics_debug_renderer.hpp#L63-L68

however that data is destructed the moment CreateTriangleBatch returns. You need to create a copy.

**Reply by cuppajoeman** [reply]
That was right on, when I copy the data instead of using the pointer, it works well: 


https://github.com/jrouwe/JoltPhysics/assets/19198104/3ee0b198-cece-4b99-a1b1-d7c7efd72651

**Comment by MrOnlineCoder** [comment]
`My game doesn't use a LOD system yet, I don't fully understand how I'm supposed to use the LOD array.`
I don't know the correct way there, but I used just the first LOD passed to DrawGeomtetry and it worked fine for my purposes of just being able to visually verify the shapes/bodies are matching my world geometry:
```cpp
auto lod = inGeometry.GetPtr()->mLODs[0];
```

**Reply by cuppajoeman** [reply]
@jrouwe I changed my code to instead just use the first LOD value and it seems to work fine, can you explain why this happens and we don't have to iterate through the LOD array?

**Reply by jrouwe** [reply]
The LOD array is only needed in case you can't render the scene at the target FPS anymore (i.e. if it consists of hundreds of thousands of triangles), if you don't have that then picking the first LOD is fine. 

You should know where your camera is and you get the bounding box as part of the DrawGeometry function so you can calculate the distance from camera to the box and pick the lowest LOD whose distance is smaller than the calculated distance. The lod scale also factors into this (this is for when an object has been scaled) the complete formula is here:

https://github.com/jrouwe/JoltPhysics/blob/94124f08da0a3a8d26f70e4937e058b5b21aa860/TestFramework/Renderer/DebugRendererImp.cpp#L339

**Reply by cuppajoeman** [reply]
Ok, that makes sense, if the renderer starts lagging, I'm going to look more into this

**Comment by cuppajoeman** [comment]
Hey @jrouwe I was just going back and making sure my project compiles correctly, when I compile my source code (which contains an implementation of the `DebugRenderer`. Here's the [header file](https://github.com/opengl-toolbox/physics_debug_renderer/blob/ce810f7c6ff11ce478ab3ca9b62e03a3bd9608b3/physics_debug_renderer.hpp) and the [source](https://github.com/opengl-toolbox/physics_debug_renderer/blob/ce810f7c6ff11ce478ab3ca9b62e03a3bd9608b3/physics_debug_renderer.cpp)) using either `Release` or `Debug` everything compiles and works properly.

If I don't specify the build type I get the following linker error:
```
[ 96%] Linking CXX executable mwe_physics_world_with_character
/usr/bin/ld: CMakeFiles/mwe_physics_world_with_character.dir/main.cpp.o: in function `PhysicsDebugRenderer::~PhysicsDebugRenderer()':
main.cpp:(.text._ZN20PhysicsDebugRendererD2Ev[_ZN20PhysicsDebugRendererD5Ev]+0x22): undefined reference to `JPH::DebugRenderer::~DebugRenderer()'
/usr/bin/ld: CMakeFiles/mwe_physics_world_with_character.dir/graphics/physics_debug_renderer/physics_debug_renderer.cpp.o: in function `PhysicsDebugRenderer::PhysicsDebugRenderer()':
physics_debug_renderer.cpp:(.text+0x15): undefined reference to `JPH::DebugRenderer::DebugRenderer()'
/usr/bin/ld: physics_debug_renderer.cpp:(.text+0xa6): undefined reference to `JPH::DebugRenderer::Initialize()'
/usr/bin/ld: physics_debug_renderer.cpp:(.text+0xb7): undefined reference to `JPH::DebugRenderer::~DebugRenderer()'
/usr/bin/ld: CMakeFiles/mwe_physics_world_with_character.dir/graphics/physics_debug_renderer/physics_debug_renderer.cpp.o:(.data.rel.ro._ZTI20PhysicsDebugRenderer[_ZTI20PhysicsDebugRenderer]+0x10): undefined reference to `typeinfo for JPH::DebugRenderer'
/usr/bin/ld: CMakeFiles/mwe_physics_world_with_character.dir/graphics/graphics.cpp.o: in function `JPH::PhysicsSystem::DrawBodies(JPH::BodyManager::DrawSettings const&, JPH::DebugRenderer*, JPH::BodyDrawFilter const*)':
graphics.cpp:(.text._ZN3JPH13PhysicsSystem10DrawBodiesERKNS_11BodyManager12DrawSettingsEPNS_13DebugRendererEPKNS_14BodyDrawFilterE[_ZN3JPH13PhysicsSystem10DrawBodiesERKNS_11BodyManager12DrawSettingsEPNS_13DebugRendererEPKNS_14BodyDrawFilterE]+0x44): undefined reference to `JPH::BodyManager::Draw(JPH::BodyManager::DrawSettings const&, JPH::PhysicsSettings const&, JPH::DebugRenderer*, JPH::BodyDrawFilter const*)'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/mwe_physics_world_with_character.dir/build.make:265: mwe_physics_world_with_character] Error 1
make[2]: *** [CMakeFiles/Makefile2:222: CMakeFiles/mwe_physics_world_with_character.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:229: CMakeFiles/mwe_physics_world_with_character.dir/rule] Error 2
make: *** [Makefile:169: mwe_physics_world_with_character] Error 2
```

I tried figuring out how to fix this, and I found that in your implementation you have this line: 

https://github.com/jrouwe/JoltPhysics/blob/5c9d3b011ab76cdaa05c9e1b1e699f716bd1ddc4/TestFramework/Renderer/DebugRendererImp.cpp#L16

where you include the source file rather than the header file under some specific conditions. When I include that source file (`DebugRenderer.cpp`) it reduces the number of errors: 
```
[ 95%] Linking CXX executable mwe_physics_world_with_character
/usr/bin/ld: CMakeFiles/mwe_physics_world_with_character.dir/graphics/graphics.cpp.o: in function `JPH::PhysicsSystem::DrawBodies(JPH::BodyManager::DrawSettings const&, JPH::DebugRenderer*, JPH::BodyDrawFilter const*)':
graphics.cpp:(.text._ZN3JPH13PhysicsSystem10DrawBodiesERKNS_11BodyManager12DrawSettingsEPNS_13DebugRendererEPKNS_14BodyDrawFilterE[_ZN3JPH13PhysicsSystem10DrawBodiesERKNS_11BodyManager12DrawSettingsEPNS_13DebugRendererEPKNS_14BodyDrawFilterE]+0x44): undefined reference to `JPH::BodyManager::Draw(JPH::BodyManager::DrawSettings const&, JPH::PhysicsSettings const&, JPH::DebugRenderer*, JPH::BodyDrawFilter const*)'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/mwe_physics_world_with_character.dir/build.make:265: mwe_physics_world_with_character] Error 1
make[2]: *** [CMakeFiles/Makefile2:222: CMakeFiles/mwe_physics_world_with_character.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:229: CMakeFiles/mwe_physics_world_with_character.dir/rule] Error 2
make: *** [Makefile:169: mwe_physics_world_with_character] Error 2
```

But I don't believe this is a good workaround. Do you have any idea what's going on here when I don't specify the built type?

**Reply by jrouwe** [reply]
The manual inclusion of DebugRenderer.cpp in DebugRendererImp.cpp is there to be able to use the debug renderer without JPH_DEBUG_RENDERER being defined inside the Jolt library and is probably not something you should try to replicate. The Samples app uses it because it has no other way of rendering stuff.

Your link error probably indicates that Jolt has been built without the JPH_DEBUG_RENDERER define and that your own code has it defined. Make sure this matches on both sides.

**Reply by cuppajoeman** [reply]
I was building jolt without specifying a build type, and that was the problem, I found out through the docs that you have to build using Debug or Release to have it work.

**Comment by cuppajoeman** [comment]
Hey @jrouwe I recently updated my jolt version from this commit: 

https://github.com/jrouwe/JoltPhysics/tree/ed970874e4d1abbdc1e056fd4ca7f2d487af8b60

to the newest commit, and now I get the following error in my previously working `physics_debug_renderer.cpp` during linking:

```
[ 75%] Linking CXX executable client
mold: error: undefined symbol: typeinfo for JPH::DebugRenderer
>>> referenced by physics_debug_renderer.cpp
>>>               CMakeFiles/client.dir/src/graphics/physics_debug_renderer/physics_debug_renderer.cpp.o:(std::_Vector_base<unsigned int, std::allocator<unsigned int> >::_Vector_impl::~_Vector_impl())
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/client.dir/build.make:883: client] Error 1
make[1]: *** [CMakeFiles/Makefile2:103: CMakeFiles/client.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

any hints on why this might have popped up between that commit and now?

For more details my debug renderer implementation is here:
* [physics_debug_renderer.hpp](https://github.com/cpp-toolbox/physics_debug_renderer/blob/58b03953545232c99418e322158f38e49a1eae3e/physics_debug_renderer.hpp)
* [physics_debug_renderer.cpp](https://github.com/cpp-toolbox/physics_debug_renderer/blob/58b03953545232c99418e322158f38e49a1eae3e/physics_debug_renderer.cpp)

**Comment by jrouwe** [comment]
Jolt by default compiles without C++ RTTI, your project is compiling with RTTI. Either turn it off for your own project or turn it on for Jolt using the CPP_RTTI_ENABLED cmake option.

**Reply by cuppajoeman** [reply]
Just wanted to mention that this worked.

**Comment by cuppajoeman** [comment]
Hey, so I have a new rendering implementation and wanted to try out `DebugRendererSimple` to test it out. (I'm using jolt 5.2) Here's my current implementation:

```cpp
class PhysicsDebugRenderer : public JPH::DebugRendererSimple {
  private:
    CW_V_TransformationWithColoredVertexShaderRenderer &renderer;

    glm::vec3 colorToVec3(JPH::ColorArg color) {
        return glm::vec3(color.r / 255.0f, color.g / 255.0f, color.b / 255.0f);
    }

    glm::vec3 rvec3ToVec3(JPH::RVec3Arg vec) {
        return glm::vec3(static_cast<float>(vec.GetX()), static_cast<float>(vec.GetY()),
                         static_cast<float>(vec.GetZ()));
    }

  public:
    explicit PhysicsDebugRenderer(CW_V_TransformationWithColoredVertexShaderRenderer &renderer) : renderer(renderer) {
        std::cout << "in constructor" << std::endl;
    }

    virtual void DrawLine(JPH::RVec3Arg inFrom, JPH::RVec3Arg inTo, JPH::ColorArg inColor) override {
      ...
    }

    virtual void DrawTriangle(JPH::RVec3Arg inV1, JPH::RVec3Arg inV2, JPH::RVec3Arg inV3, JPH::ColorArg inColor,
                              ECastShadow inCastShadow) override {
        std::cout << "about to draw triangle" << std::endl;

        std::vector<glm::vec3> positions = {rvec3ToVec3(inV1), rvec3ToVec3(inV2), rvec3ToVec3(inV3)};

        glm::vec3 color = colorToVec3(inColor);
        std::vector<glm::vec3> colors = {color, color, color};

        std::vector<unsigned int> indices = {0, 1, 2};

        renderer.draw(indices, positions, colors);
    }

    virtual void DrawText3D(JPH::RVec3Arg inPosition, const std::string_view &inString, JPH::ColorArg inColor,
                            float inHeight) override {

        std::cout << "about to draw text" << std::endl;
        // Not implemented as requested
    }
};
```

I'm getting a segfault with this even before "in constructor" can occur. The backtrace looks like this: 
```
Thread 1 "cmr_camera" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00005555562054de in JPH::STLAllocator<JPH::DebugRenderer::Vertex>::reallocate<true, void> (this=0x7ffffffda280, inOldPointer=0x0, inOldSize=0, inNewSize=4)
    at /home/ccn/.conan2/p/b/joltp6fa058938bc00/b/src/Build/../Jolt/Core/STLAllocator.h:66
#2  0x00005555562037ea in JPH::Array<JPH::DebugRenderer::Vertex, JPH::STLAllocator<JPH::DebugRenderer::Vertex> >::reallocate (this=0x7ffffffda280, inNewCapacity=4)
    at /home/ccn/.conan2/p/b/joltp6fa058938bc00/b/src/Build/../Jolt/Core/Array.h:84
#3  0x0000555556202731 in JPH::Array<JPH::DebugRenderer::Vertex, JPH::STLAllocator<JPH::DebugRenderer::Vertex> >::reserve (this=0x7ffffffda280, inNewSize=4)
    at /home/ccn/.conan2/p/b/joltp6fa058938bc00/b/src/Build/../Jolt/Core/Array.h:114
#4  0x0000555556201ead in JPH::Array<JPH::DebugRenderer::Vertex, JPH::STLAllocator<JPH::DebugRenderer::Vertex> >::resize (this=0x7ffffffda280, inNewSize=4)
    at /home/ccn/.conan2/p/b/joltp6fa058938bc00/b/src/Build/../Jolt/Core/Array.h:121
#5  0x00005555561f9ca1 in JPH::DebugRenderer::CreateQuad (this=0x7ffffffdcda0, ioIndices=..., ioVertices=..., inV1=..., inV2=..., inV3=..., inV4=...)
    at /home/ccn/.conan2/p/b/joltp6fa058938bc00/b/src/Jolt/Renderer/DebugRenderer.cpp:429
#6  0x00005555561fa264 in JPH::DebugRenderer::Initialize (this=0x7ffffffdcda0) at /home/ccn/.conan2/p/b/joltp6fa058938bc00/b/src/Jolt/Renderer/DebugRenderer.cpp:481
#7  0x00005555562087e9 in JPH::DebugRendererSimple::DebugRendererSimple (this=0x7ffffffdcda0)
    at /home/ccn/.conan2/p/b/joltp6fa058938bc00/b/src/Jolt/Renderer/DebugRendererSimple.cpp:15
#8  0x0000555555792bbf in PhysicsDebugRenderer::PhysicsDebugRenderer (this=0x7ffffffdcda0, renderer=...)
    at /home/ccn/projects/cpp_toolbox_organization/jolt_physics_sandbox/src/graphics/physics_debug_renderer/physics_debug_renderer.hpp:41
#9  0x000055555578fdb9 in main () at /home/ccn/projects/cpp_toolbox_organization/jolt_physics_sandbox/src/main.cpp:68
(gdb) 
```
I'm not sure what I'm missing with what I've done, does anything stand out? @jrouwe

**Reply by jrouwe** [reply]
Please mind my time and do your own investigations before asking trivial questions: On STLAllocator.h line 66 there's a function call to `Reallocate` which jumps to a nullptr, probably because you haven't implemented the `Reallocate` function or register it too late.

**Reply by cuppajoeman** [reply]
Hi Jorrit,

I'm sorry if my questions have felt trivial or like I'm wasting your time, that's not my intention.

The solution here is a trivial mistake on my part: I hadn’t initialized the physics system before setting up the debug renderer, which caused the crash. 

But as someone new to Jolt, finding the solution was not trivial to me. The error I got: a nullptr dereference in Reallocate didn’t point me in the right direction. I ended up spending a lot of time digging through the source code to try and understand why sReallocate would be null and couldn't think of a logical reason.  It might seem obvious in hindsight, but I really did spend a lot of time trying to figure it out before reaching out.
